File: samples/C/sample2.c
#include <stdio.h>
#include <string.h>
#include "vec3d.h"
#include "sample_base.h"
#include "abstract.h"

/* Mode CONST */
#define GO_TO_WAITING_POINT   201
#define WAIT_UNTIL_MESSAGE    202
#define GO_TO_SWITCH          203
#define BACK_TO_WAITING_POINT 204

/* Event CONST */
#define ARRIVAL_WAITING_POINT_EVENT 201
#define ARRIVAL_SWITCH_EVENT        202

/* Location of waiting point. */
const vec3d waitingPoint = {0,0,30};

/* The mode of this car. */
int mode = GO_TO_WAITING_POINT;

void make_events() {
  vec3d tmpV;
  struct event e;
  make_events_basic();

  // car has arrived at the waiting point?
  v3sub(&waitingPoint,&loc,&tmpV);
  if (v3length(&tmpV)<1.0) {
    e.id = ARRIVAL_WAITING_POINT_EVENT;
    process_event(&e);
  }

  // car has arrived at the switch?
  v3sub(&switch1,&loc,&tmpV);
  if (v3length(&tmpV)<1.0) {
    e.id = ARRIVAL_SWITCH_EVENT;
    process_event(&e);
  }
}

void process_event(struct event *e) {
  if (mode==GO_TO_WAITING_POINT && e->id==ARRIVAL_WAITING_POINT_EVENT) {
    mode = WAIT_UNTIL_MESSAGE;
  } else if (mode==GO_TO_SWITCH && e->id==ARRIVAL_SWITCH_EVENT) {
    mode = WAIT_UNTIL_MESSAGE;
  } else if (mode==WAIT_UNTIL_MESSAGE && e->id==MESSAGE_EVENT) {
printf("sample2:receiveMessage: %s\n",e->message);
    if (strcmp(e->message,"wait")==0) {
      mode = BACK_TO_WAITING_POINT;
    } else if (strcmp(e->message,"pushSwitch")==0) {
      mode = GO_TO_SWITCH;
    }
  } else if (mode==BACK_TO_WAITING_POINT && e->id==ARRIVAL_WAITING_POINT_EVENT) {
    mode = WAIT_UNTIL_MESSAGE;
  } else {
    //printf("Unprocessed event: id=%d\n",e->id);
  }
}

void go_to_waiting_point() {
  vec3d v;
  if (check_all_conflict(&loc,&waitingPoint,NULL)) {
printf("GAHA:CONFLICT\n");
    v3sub(&waitingPoint,&loc,&v);
    v3simpleRotateY(&v,45);
    v3add(&v,&loc,&v);
    go_to_destination(&v);
  } else {
    go_to_destination(&waitingPoint);
  }
}

void wait_until_message() {
  stop_car();
}

void go_to_switch() {
  go_to_destination(&switch1);
}

void back_to_waiting_point() {
  back_to_destination(&waitingPoint);
}

void move() {
//printf("mode=%d\n",mode);
  switch(mode) {
  case GO_TO_WAITING_POINT:   go_to_waiting_point();   break;
  case WAIT_UNTIL_MESSAGE:    wait_until_message();    break;
  case GO_TO_SWITCH:          go_to_switch();          break;
  case BACK_TO_WAITING_POINT: back_to_waiting_point(); break;
  default: printf("sample2: Unknown mode?%d\n",mode);
  }
}

int main(void) {
  printf("start blue car!\n");
  init_car(20000);
  start();
}

File: samples/Java/Sample2.java
/*
 * This enum represents mods of the car (Sample2).
 * These modes correspond to states of FSM (Finite State Machine).
 */
enum S2Mode {
    GO_TO_WAITING_POINT,
    WAIT_UNTIL_MESSAGE,
    GO_TO_SWITCH,
    BACK_TO_WAITING_POINT
}

/*
 * The following classes represent various events.
 * These events are created in stateCheck() method
 * and passed to processEvent() method.
 */
class ArrivalWaitingPointEvent extends Event {}
class ArrivalSwitchEvent extends Event {}

/*
 * Sample2 is a program which controls the blue car in the
 * simulation environment. This car operates the elevator
 * on demand of the red car. Basic functions are implemented
 * in the SampleBase class which is extended by this Sample2 class.
 */
public class Sample2 extends SampleBase {
    // Location of waiting point.
    static final Vector waitingPoint = new Vector(0,0,30);

    // The mode of this car
    S2Mode mode;

    /*
     * The constructor of Sample2. super(20000) means
     * that the blue car is controled through port 20000
     * (computer networking).
     */
    public Sample2() {
        super(20000);
        mode = S2Mode.GO_TO_WAITING_POINT;
    }

    /*
     * Check the situations of this car and create some
     * events then call processEvent() method. To process
     * general events, this method call super.stateCheck()
     * at first.
     */
    @Override
    protected void stateCheck() { 
        super.stateCheck();
        Vector tmpV = new Vector();

        // car has arrived at the waiting point?
        tmpV.sub(waitingPoint,loc);
        if (tmpV.length()<1.0)
            processEvent(new ArrivalWaitingPointEvent());

        // car has arrived at the switch?
        tmpV.sub(switch1,loc);
        if (tmpV.length()<1.0)
            processEvent(new ArrivalSwitchEvent());
    }

    /*
     * Decide the next mode in consideration of the previous
     * mode and the given event. The process is based on FSM
     * (finite state machine). This method implements a strategy
     * of this car.
     */
    @Override
    protected void processEvent(Event e) {
        if ((mode==S2Mode.GO_TO_WAITING_POINT)
          &&(e instanceof ArrivalWaitingPointEvent)) {
            mode = S2Mode.WAIT_UNTIL_MESSAGE;
        } else if ((mode==S2Mode.GO_TO_SWITCH)
          &&(e instanceof ArrivalSwitchEvent)) {
            mode = S2Mode.WAIT_UNTIL_MESSAGE;
        } else if ((mode==S2Mode.WAIT_UNTIL_MESSAGE)
                 &&(e instanceof MessageEvent)) {
            String message = ((MessageEvent)e).message;
System.out.println("Sample2:reciveMessage: "+message);
            if (message.equals("wait")) {
                mode = S2Mode.BACK_TO_WAITING_POINT;
            } else if (message.equals("pushSwitch")) {
                mode = S2Mode.GO_TO_SWITCH;
            }
        } else if ((mode==S2Mode.BACK_TO_WAITING_POINT)
                 &&(e instanceof ArrivalWaitingPointEvent)) {
            mode = S2Mode.WAIT_UNTIL_MESSAGE;
        } else {
            //System.out.println("Unprocessed event: "+e.getClass().getName());
        }
    }

    /*
     * Control the car in accordance with the mode of the car.
     */
    @Override
    protected void move() {
        switch(mode) {
        case GO_TO_WAITING_POINT: goToWaitingPoint(); break;
        case WAIT_UNTIL_MESSAGE: waitUntilMessage(); break;
        case GO_TO_SWITCH: goToSwitch(); break;
        case BACK_TO_WAITING_POINT: backToWaitingPoint(); break;
        default: System.out.println("Sample2: Unknown mode?"+mode);
        }
    }

    // The following methods implement processes for each mode.

    void goToWaitingPoint() {
        goToDestination(waitingPoint);
    }
    void waitUntilMessage() {
        stopCar();
    }
    void goToSwitch() {
        goToDestination(switch1);
    }

    void backToWaitingPoint() {
        backToDestination(waitingPoint);
    }

    /*
     * The start point of Sample2.
     */
    public static void main(String args[]) {
        Sample2 s2 = new Sample2();
        s2.start();
    }
}
